octobot_trading/personal_data/portfolios/history/historical_portfolio_value_manager.py

```diff
@@ -0,0 +1,267 @@
+#
+#  This library is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+#  Lesser General Public License for more details.
+#
+#  You should have received a copy of the GNU Lesser General Public
+#  License along with this library.
+import sortedcontainers
+
+import octobot_commons.logging as logging
+import octobot_commons.databases as databases
+import octobot_commons.constants as commons_constants
+import octobot_commons.enums as commons_enums
+import octobot_commons.errors as commons_errors
+import octobot_commons.symbol_util as symbol_util
+
+import octobot_trading.util as util
+import octobot_trading.errors as errors
+import octobot_trading.constants as constants
+import octobot_trading.personal_data.portfolios.history.historical_asset_value as historical_asset_value
+
+
+class HistoricalPortfolioValueManager(util.Initializable):
+    """
+    HistoricalPortfolioValueManager stores and make the portfolio value through time as HistoricalAssetValue
+    """
+    TABLE_NAME = "historical_portfolio_value"
+    DATA_SOURCE_KEY = "data_source"
+    DATA_VERSION_KEY = "data_version"
+    DEFAULT_DATA_SOURCE = "portfolio_value_holder"  # for later versions: consolidate data with transaction history
+    DEFAULT_DATA_VERSION = 1
+    MIN_TIME_FRAME_RELEVANCY_SECONDS = 29   # less than 30s to support 1m time frames
+    TIME_FRAME_RELEVANCY_TIME_RATIO = 0.45   # allow for 45% error in timestamp rounding (only used after the ideal
+    # time passed to avoid using past data: use future data at worse). use high value to still take value if users
+    # start it in the 1st part of the day for example
+    MAX_HISTORY_SIZE = 250000
+
+    def __init__(self, portfolio_manager, data_source=None, version=None):
+        super().__init__()
+        self.portfolio_manager = portfolio_manager
+        self.logger = logging.get_logger(f"{self.__class__.__name__}"
+                                         f"[{self.portfolio_manager.exchange_manager.exchange_name}]")
+        self._portfolio_type_suffix = self._get_portfolio_type_suffix()
+        self.saved_time_frames = self.portfolio_manager.exchange_manager.config.get(
+            commons_constants.CONFIG_SAVED_HISTORICAL_TIMEFRAMES,
+            constants.DEFAULT_SAVED_HISTORICAL_TIMEFRAMES
+        )
+        self.data_source = data_source or self.__class__.DEFAULT_DATA_SOURCE
+        self.version = version or self.__class__.DEFAULT_DATA_VERSION
+
+        self.max_history_size = self.__class__.MAX_HISTORY_SIZE
+        self.historical_portfolio_value = sortedcontainers.SortedDict()
+        try:
+            self.run_dbs_identifier = util.get_run_databases_identifier(self.portfolio_manager.exchange_manager)
+        except commons_errors.ConfigTradingError:
+            # can't save data without an activated trading mode
+            self.run_dbs_identifier = None
```

Comment:
I think we should create a proxy to prevent doing this each time we need to prepare writing into a Database

Reply:
Right, now we use `storage`